### DISCUSSION TITLE
[TCP Feature] Issue using OtTcpSendByExtension

### DISCUSSION THREAD

**Discussion by estelleOl** [openthread/openthread#8903] [Q&A] [answered]
Hello,

I am trying to use to `OtTcpSendByExtension` method to send a TCP message in to parts.

I have 2 boards a leader and a child, the child is sending the first part of the message using OtTcpSendByReference, then change the size message using OtTcpSendByExtension and finally send the second part of the message with OtTcpSendByReference.

The result I observe is the "msg1" + "/0" + "/0" + "msg2", I was expecting "msg1" + "msg2" instead. What is the message I should expect ?

I join the board2 code under board 1 is just waiting to receive by reference or by extension.

```
otLinkedBuffer    linkedBuffer    = {0};
char dataBoard2[] = "Board2";
char dataByExtension[] = " | Adding message by extension ";
uint8_t                 lengthExtension = strlen(dataByExtension);
uint8_t                 lengthMsg = strlen(dataBoard2);

 /* Send data referenced by the linked buffer pointed to */
 linkedBuffer.mData = (const uint8_t*)dataBoard2;
 linkedBuffer.mLength = lengthMsg ;
 status = OtTcpSendByReference(&tcpEndPoint, &linkedBuffer,OT_TCP_SEND_MORE_TO_COME);
         
 if(status != OT_ERROR_NONE){
 return;
 }

status = OtTcpSendByExtension(&tcpEndPoint, 
lengthExtension, OT_TCP_SEND_MORE_TO_COME);
          
if(status != OT_ERROR_NONE){
return;
}

/* Send data referenced by the linked buffer pointed to */
linkedBuffer.mData = (const uint8_t*)dataByExtension;
linkedBuffer.mLength = lengthExtension;
status = OtTcpSendByReference(&tcpEndPoint, &linkedBuffer, 0);
          
if(status != OT_ERROR_NONE){
return;
 }
```

The message I get on Borad 1 is "Board2\0\0 | Adding message by extension "

Is it what I am supposed to have ? I wanted to have "Board2 |  Adding message by extension "

Regards,

Estelle

**Comment by abtink** [comment, marked as answer]
Check the documentation of [`otTcpSendByExtension()`](https://github.com/openthread/openthread/blob/e8f3ec0fd3356d0c52b1a6f444e63b5c86abbbdb/include/openthread/tcp.h#L455-L469).

Here are some suggestions:
- Make sure you have a single array buffer that can fit both messages , i.e.,can contain  `"Board 2 | Adding message by extnsion"`.
  - Make sure both array the `otLinkedBuffer` persist (should not be be local stack variables).
- Use this array when you prepare the `otLinkedBuffer` and copy the first message into buffer.
- Call `otTcpSendByReference()` which takes ownership of the `otLinkBuffer` (so we should not change it anymore).
- To send next message, you can write second message into the array buffer at the end (after the last message).
   -  Important: Do not change the `otLinkBuffer`'s `mLength` field (it is owned by TCP stack now).
- Then call `otTcpSendByExtension()` indicating number of new bytes you added to the previous buffer as input `aNumBytes`.

cc: @samkumar.

**Reply by estelleOl** [reply]
Thank you for your response it has helped me to solve my issue.
